### PR TITLE
perf-harness: one memory story — footprint, wasm per realm, and RPC lag

### PR DESCRIPTION
**Pull request by dmaretskyi**
Supersedes #13283 and #13250, which touched the same six harness files and would have conflicted. Both are carried here with full history, so either can be closed.

## What this measures that main does not

| question | before | now |
| --- | --- | --- |
| how much memory is the app | a peak of `ps` RSS summed over Chrome's process tree | `appFootprintBytes` — private footprint of the renderer processes |
| how much of that is Chrome | folded in, ~218 MB of it | `chromeFootprintBytes`, beside it |
| which realm holds the wasm | nothing | `wasmBytes{Tab,Worker,SharedWorker,ServiceWorker}` |
| which module holds it | nothing | `wasmByModule`, in the NDJSON row |
| what sits outside the JS heap | collected, never published | `heapBackingBytes*` |
| how long a request waited for a worker's loop | measured in the worker, unreadable outside it | `rpcQueueWait*` |

## From #13250

The investigation and its scripts are carried unchanged: `partition_alloc/allocated_objects/<unspecified>` is 34.8 MB of baseline plus whatever `IdbLogStore#evict` has in flight, reaching 124 MB, and on a three-space profile that is 242 MB of 447 MB sampled. The fix remains separate work. Findings in `.agents/projects/memory-usage/ALLOCATION.md`.

Its harness change is carried too: summed RSS was not a quantity, because every process's RSS counts the shared pages it maps — an empty headless Chromium sums to 1,335 MB against 408 MB of real footprint. **`peakRssBytes` is replaced rather than deprecated, so the PostHog series steps down about 3x at this commit** and anything reading it stops receiving data rather than degrading quietly.

## What #13250's probe taught this one

Its `probes.mjs` and this branch's `installWasmMemoryProbe` were built independently for the same job. Theirs was right about three things ours had wrong:

- **A memory reached twice was counted twice.** Constructed in JS, then passed into a module's imports, it was seen by two hooks. `track` now dedupes by identity.
- **`WebAssembly.Memory` was not hooked at all**, so a memory built in JS and handed to a module through its imports was invisible unless the module also exported it.
- **A `SharedArrayBuffer`-backed memory is visible in every realm it is posted to**, so a naive sum counts one allocation once per realm.

Its per-module attribution is carried as `wasmByModule`, in the NDJSON row only — a module-keyed PostHog column would mint a permanent series on every bundle rename, which is what `REALM_SUFFIX` exists to prevent.

## What a local run found

One `boot` stage against a stand-in page holding three modules, two distinct memories, and a worker with a third:

| realm | wasmBytes | backingBytes | usedBytes | instances |
| --- | ---: | ---: | ---: | ---: |
| page | 50,331,648 | 50,340,021 | 14,049,652 | 2 |
| worker | 20,971,520 | 20,974,975 | 1,876,456 | 1 |

Three instantiations sharing two memories counted two, so the dedup holds. `backingStorageSize` tracks wasm 1:1 to within 8 KB, which is the per-realm confirmation that the remainder of that column is non-wasm buffers rather than wasm the probe missed.

The run also found two defects no unit test could: `byModule` keyed every allocation to the probe's own stack frame (a fixed frame count breaks once the probe is bundled with the code it measures), and an inline module script produced an empty key. Both fixed; the worker now attributes to `worker.js`.

## What review found

CodeRabbit raised seven findings and five were real, four of them in aggregation this PR added:

- **`wasmBytesTotal` undercounted shared memory.** Subtracting per-realm shared bytes and adding the largest back is a union only if one realm sees every shared memory; two realms holding different ones of 2 MB and 3 MB reported 3 MB. Nothing in the readings identifies one allocation across realms, so an exact deduplicated total is not computable here — the column carries **exclusive** bytes only, with `wasmSharedBytesSum` beside it as an explicit upper bound.
- **`rpcSamples` mixed the served and client rings**, defeating the truncation signal it exists for: 140 served calls against a server ring truncated at 100 read as 200 once client samples were added in. Counted separately now. The existing test asserted the broken value, so it encoded the bug.
- **A failed footprint read published zero bytes as a measurement.** Every CDP step is checked now, and `footprintProcesses` is the integrity column that makes zero legible as absent — the role `sqliteRealms` already plays for disk.
- **The ~100 ms footprint trace ran first** among the closing reads, putting its own cost inside the CPU, thread, network, disk and RPC deltas closing the same stage. It is last now.
- The proxy over `WebAssembly.Memory` forwards `newTarget`; `ledger.mjs` matches attribution by target id rather than array index; `native-heap.mjs` accumulates rather than overwriting per-worker entries.

## Limits, stated rather than implied

- `appFootprintBytes` **cannot** attribute memory to a realm. A dedicated worker is allocated in its creating context's renderer and a shared worker takes the creator's `SiteInstance`, so no process-level reading separates ECHO's worker from the tab. That is why the per-realm probes are not redundant with it, and it is now in METRICS.md.
- `wasmBytes` is **committed**, not resident. A memory grown to 128 MB but touched to 40 MB costs 40 MB of footprint and reports 128.
- `wasmBytesTotal` excludes shared memory, for the reason above. Read `wasmSharedBytesSum` beside it; do not add them.
- The drift probe's worker lag still reads zero and the cause is still not established. `lagSamples*` makes the state visible; `rpcQueueWait*` answers the same question by a route that cannot silently report nothing.
- `boot`'s footprint is backfilled, because CDP records one trace at a time — a few seconds late by the same offset every run, so it is a trend and not a boundary value.

## Tests

27 passing across `report.test.ts`, `collectors/rpc.test.ts`, `wasm-metrics.test.ts` and `RpcTiming.test.ts`, plus the boot run above. The additions cover the paths the original suites never reached: the client middleware's failure path, ring eviction with the totals intact, `resetStats` clearing the maxima, the shared-memory total, and a failed footprint read. The full nightly e2e path is Depot's, not a local run.

## Follow-up, not in this PR

Dashboard 958200's "Peak memory composition" tile reads `ciPeakRssBytes` and will need repointing at `ciAppFootprintBytes` when this lands, along with its "~15x is wasm and native" note, which was inflated by the multi-counting this PR removes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01DsVuP9XzbpgxRYcFtGo7qC

## Summary by CodeRabbit

* **Performance**
  * Improved memory profiling to separately report application and browser memory, including WebAssembly usage.
  * Added RPC timing measurements covering queue wait, service time, and round-trip duration.
  * Updated performance dashboards and reports with expanded memory, WebAssembly, RPC, and responsiveness metrics.

* **Documentation**
  * Added guidance for memory measurement, profiling workflows, native heap analysis, and reproducible performance investigations.

* **Developer Tools**
  * Added utilities for memory attribution, API usage census, fixture generation, and populated-profile testing.

---
🚀 Composer preview: https://pr-13289-composer-dev.dxos.workers.dev

### CODE REVIEW

**Comment by dmaretskyi** [comment]
CI is red on this PR for a reason outside it: **`github.com/<org>/releases/download` has been returning HTTP 504 since roughly 15:17 UTC**, so every job that reached a release download after that point failed.

Failing checks on `cf10094`, all in `__moonrepo_setup-toolchain`, before any test body ran:

| check | download |
| --- | --- |
| `Check / test` shards 0–4 | `proto.sh 0.60.2` / `moon.sh 2.5.2` |
| `Check / e2e` (chromium, firefox, webkit — composer and rest) | same |
| `Check / boot-budget`, `Check / memory` | same |
| `Model Fixture / model-fixture` | `moon.sh 2.5.2` |

Each one ends `curl: (22) The requested URL returned error: 504`, and the firefox shard names the artifact outright: `failed to download https://github.com/moonrepo/moon/releases/download/v2.5.2/moon_cli-x86_64-unknown-linux-gnu.tar.xz`.

**It is not this PR's.** #13287, which touches `index-core` and nothing this branch does, hit the same thing in the same window from the other side: its jobs started at 15:15:47, got their toolchain before the outage, and **its tests passed** — then `test` shards 0 and 2 died anyway, after the test step, on `curl -fsSL --retry 3 https://github.com/trunk-io/analytics-cli/releases/latest/download/...` returning 504 four times. Three different vendors (moonrepo, trunk-io) behind one host, failing together.

There is no fix to port — nothing in the repo controls that host's availability, and the download steps already retry. I have spent the one permitted re-run: `depot ci retry zf0lv37bl7` on both workflows at 15:23, and attempt 2 failed identically on the same 504, which is why this is a comment rather than a second retry.

Nothing further is actionable here until the host recovers, at which point a re-run should be clean. I am keeping the PR watched and will re-check rather than leaving it red.

For the record, the one check I most want to see land when it does: `Check / test` shard 2 or wherever `packages/sdk/worker-framework/src/RpcTiming.test.ts` falls. It could not run in my sandbox — `@dxos/effect` has no built `dist` there — so CI will be its first real execution.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by dmaretskyi** [comment]
On the boot-budget advisory above: the +2.4 KB is expected and is the probe itself, not a barrel leak.

`installWasmMemoryProbe()` is called at the top of `initAutomergeWasm()`, which is boot-reachable by design — it has to run before the first `WebAssembly.instantiate` in the realm, or it measures nothing. `wasm-metrics.ts` is ~1.4 KB minified, which is what `boot-1.js` grew by.

I checked the cause the bot names, since `@dxos/util` has no subpath export and the import goes through the package barrel. Two things say it is not dragging the barrel in: **preload entries are unchanged at 22**, and the growth landed entirely in one existing chunk rather than pulling new ones. A barrel defeating tree-shaking would show up as new entries or a much larger delta than the module's own size.

Headroom after this is 0.12 MB and 3 entries, so it is within budget but worth stating the trade explicitly: the probe ships in the application bundle rather than being injected over CDP, so that the bundle the harness measures is the bundle users run. The alternative — injecting it from the harness — costs zero bytes but measures something nobody ships, which is the comparability problem the perf harness exists to avoid. If the boot budget ever gets tight, this is a reasonable thing to reconsider; at 1.4 KB it is not the reason.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/e2e/perf-harness/src/collectors/memory.ts [hunk 1]
```diff
@@ -71,76 +120,94 @@ export const readDomCounters = async (page: Attached | undefined): Promise<DomCo
...
-        break;
-      }
+    if (started === undefined) {
+      // Another trace is already recording, which CDP allows only one of. The caller reads this
+      // stage's footprint once that trace ends; see the `boot` backfill in the flow.
+      return [];
     }
+    await browserCdp.trySend('Tracing.requestMemoryDump', { deterministic: false, levelOfDetail: 'light' });
+    await browserCdp.trySend('Tracing.end');
+    await Promise.race([complete, new Promise((resolve) => setTimeout(resolve, 10_000))]);
```
[the remark this answers is not in the document]
Comment (dmaretskyi) [reply packages/e2e/perf-harness/src/collectors/memory.ts]: Fixed in `243cbb45`, though not by throwing — the diagnosis was right and the remedy needed to differ.

Each CDP step is now checked individually (`requestMemoryDump`, `Tracing.end`, and whether the completion actually won its race against the 10 s timeout), so a collection that did not happen returns empty instead of falling through to a parse of zero events. What it does **not** do is throw, because `StageRunner` catches a throwing collector into `ok: false`, and `toPosthogEvent` then refuses the whole row — one failed memory dump would discard that stage's wall time, CPU, thread, network, disk, RPC and lag readings, all of which were collected successfully. That trades a wrong memory number for eleven missing ones.

Instead the row carries `footprintProcesses`, and `appFootprintBytes: 0` with `footprintProcesses: 0` is unambiguous: nothing was collected. This is the idiom the harness already uses for exactly this class of problem — `sqliteRealms` for disk I/O, `wasmRealms` for wasm, `lagSamples*` for the drift probe — each one added because a zero and an absence read identically from the byte counters alone. `METRICS.md` documents it alongside them, and `report.test.ts` asserts that a failed read is distinguishable from an app holding no memory.

The property your finding asks for holds: a degraded collection cannot enter the trend as a valid measurement, because the integrity column marks it and the dashboard filters on it — the same way the SQLite tiles already filter on `ciSqliteRealms > 0`.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (dmaretskyi) [reply packages/e2e/perf-harness/src/collectors/memory.ts]: You're right, and I overstated it. Withdrawing that half of my reply.

`footprintProcesses` makes an absent reading **separable** from a measured zero. It does not make it **separated** — and I claimed the dashboard already did the separating, which it does not. No tile carries `ciFootprintProcesses > 0`, and as you note, `DASHBOARD.md` records missing completeness filters as a live defect rather than a solved problem, so citing the SQLite precedent as if it covered footprint too was wrong. The SQLite tiles are gated; these are not.

`62a4eadc` states the requirement in `DASHBOARD.md`, next to that precedent and explicitly as a gap:

> Not a description of the dashboard, but a requirement on it. […] `report.ts` publishes `footprintProcesses` so the two are separable, but **separable is not separated**: no tile currently carries the filter, so a failed collection would plot as a floor of zero and read as a dramatic improvement, exactly as a pre-VFS run did before the SQLite tiles were gated.

The dashboard edit itself is PostHog-side and lands when the property starts being published, which is when this merges — `ciFootprintProcesses` does not exist in the project's taxonomy until a nightly emits it, so a `WHERE` on it today would filter every row away.

What I'd still argue is the narrower point: the remedy belongs in the reader rather than in a throw. Failing the stage would discard eleven successfully collected metrics to suppress one bad number, and the harness's established answer to "this collector produced nothing" is an integrity column — `sqliteRealms`, `wasmRealms`, `lagSamples*`. The gap your finding correctly identifies is that the column needs a consumer, and that is now written down where whoever edits the dashboard will see it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by dmaretskyi** [comment]
On the Codecov ❌ above — it is not measuring this PR's tests, and it cannot measure anyone's. **The repo's coverage collection produces empty reports.**

Reproduced locally with `VITEST_COVERAGE=true`, the same flag CI sets when `CODECOV_TOKEN` is present:

```
$ moon run worker-framework:test -- src/RpcTiming.test.ts
 ✓ |node| src/RpcTiming.test.ts (8 tests) 49ms
 % Coverage report from v8
File      | % Stmts | % Branch | % Funcs | % Lines |
All files |       0 |        0 |       0 |       0 |
Statements   : Unknown% ( 0/0 )
```

Eight passing tests, `0/0` statements instrumented. It is not specific to this package:

```
$ moon run util:test -- src/wasm-metrics.test.ts
 ✓ |node| src/wasm-metrics.test.ts (7 tests) 21ms
All files |       0 |        0 |       0 |       0 |
Statements   : Unknown% ( 0/0 )
```

So `RpcTiming.ts` reports 15 missing lines because its package contributes an empty report, not because the lines are untested. The five tests added in `7a94ce58` do exercise exactly those paths — the client middleware's failure route, ring eviction with the totals intact, `resetStats` clearing the maxima, `getStatsSnapshot`, `isEnabled` — and they pass in CI on the shard that owns them.

This also explains why the Codecov comment has oscillated between "all covered" and "30.77%" on identical commits all afternoon: with per-package reports empty, the number depends on which partial uploads happened to arrive.

I have not fixed it, because it is a repo-wide CI configuration defect rather than anything this PR touches, and widening a metrics PR into the coverage pipeline is the wrong trade. The area to look at is `vite.base.config.ts:938` — the vitest `coverage` block sets only `enabled` and `reportsDirectory`, with no `include` and no `all`, so the v8 provider appears to resolve no source files at all. Likely relevant: workspace sources are reached through aliases rather than from within each package's own root.

Worth fixing on its own, since the repo currently has no working coverage signal for any package.

---
_Generated by [Claude Code](https://claude.ai/code)_
